Before you turn this problem in, make sure everything runs as expected. First, **restart the kernel** (in the menubar, select Kernel$\rightarrow$Restart) and then **run all cells** (in the menubar, select Cell$\rightarrow$Run All).

Make sure you fill in any place that says `YOUR CODE HERE` or "YOUR ANSWER HERE", as well as your name and collaborators below:

In [1]:
NAME = "Caleb Andree"
COLLABORATORS = ""

---

# LAB 2:  Campaign Financial Contributions [Total: 8 points]

## Instructions

The purpose of this assignment is for you to engage with a concrete data cleaning task. This will be accomplished through a coding assignment. You will carry out this task in the present notebook, and use the notebook to document the various steps of the exercise and to answer all questions.

In particular, we will be cleaning up some campaign donation data. The DC government makes data about campaign contribution (i.e. donations) available online to the general public on its website [opendata.dc.gov](https://opendata.dc.gov). The website does not provide very much documentation about the data.

We have created a smaller sample of this data for you to explore and clean. Follow the instructions in each question, and answer the accompanying questions. Note that some of the questions will require you to write a Python function to compute the correct answer.

<div class="alert alert-info">
To grade your submission we will run your code into the autograder. The autograder will compare the value returned by your function (i.e. the student's answer) against the correct answer. The autograder code itself is not included in this notebook, so double check that your code works correctly before submitting.
</div>


## Table of Contents
<ul>
    <li><a href="#Submission-checklist">Submission checklist</a> [1 point]</li>
    <li><a href="#Q1">Question 1</a> [1 point]</li>
    <li><a href="#Q2">Question 2</a> [1 point]</li>
    <li><a href="#Q3">Question 3</a> [1 point]</li>
    <li><a href="#Q4">Question 4</a> [1 point]</li>
    <li><a href="#Q5">Question 5</a> [1 point]</li>
    <li><a href="#Q6">Question 6</a> [1 point]</li>
    <li><a href="#Q7">Question 7</a> [1 point]</li>
</ul>

## Submission checklist

**Points**: 1

Before submitting make sure that:

1. Your name is included above, plus the name of any collaborator you worked with;
2. All Markdown cells you edited are rendering correctly, especially the ones with answers.
3. You have removed any `raise NotImplementedError()` line from your code cells.

# Part A: Inspecting the data frame

The first set of questions will ask you to inspect the data. Feel free to add new code cells and make sure to comment your code with markdown cells to describe your process of reaching the answer.

## Q1

**Points:** 1

**Question:** Read the sample campaign contribution data into a pandas data frame. Inspect the columns and cells of the data set. What is most likely the column name of the variable which has the method of payment in which the donation/contribution was made? 

1. `GIS_LAST_MOD_DTTM`

2. `DATEOFRECEIPT`

3. `CONTRIBUTIONTYPE`

In [2]:
import pandas as pd

df = pd.read_csv("Campaign_Financial_Contributions_Sample.csv", index_col=0)

df.head()

OBJECTID                              COMMITTEENAME  CANDIDATENAME   
205514    205514                                 Fenty 2006   Adrian Fenty  \
169351    168351                    Muriel Bowser for Mayor  Muriel Bowser   
129869    129869                    Muriel Bowser for Mayor  Muriel Bowser   
112855    112855  Washington DC Association of Realtors PAC            NaN   
88950      88950  Washington DC Association of Realtors PAC            NaN   

        ELECTIONYEAR       CONTRIBUTORNAME   
205514          2006    Stefon  Brathwaite  \
169351          2014        Denise Rollins   
129869          2014  Antoinette Barksdale   
112855          2002        Alethea Jordan   
88950           2002        Matthew Rogers   

                                            ADDRESS CONTRIBUTORTYPE   
205514           2141 I St NW, Washington, DC 20037      Individual  \
169351    3649 11th Street NW, Washington, DC 20010      Individual   
129869        3626 Park Pl NW, Washington, DC 20010      Individual   
112855  5034 Wisconsin Ave NW, Washington, DC 20016      Individual   
88950        1506 14th St, NW, Washington, DC 20005      Individual   

       CONTRIBUTIONTYPE                                EMPLOYER   
205514            Check                                    Self  \
169351      Credit Card  US Agency of International Development   
129869            Check                          Dept of Justce   
112855            Check                                     NaN   
88950             Check                  EASE Property Services   

       EMPLOYERADDRESS  ...  Unnamed: 31 Unnamed: 32  Unnamed: 33   
205514  Washington, DC  ...          NaN         NaN          NaN  \
169351             NaN  ...          NaN         NaN          NaN   
129869             NaN  ...          NaN         NaN          NaN   
112855             NaN  ...          NaN         NaN          NaN   
88950            20005  ...          NaN         NaN          NaN   

        Unnamed: 34  Unnamed: 35  Unnamed: 36  Unnamed: 37 Unnamed: 38   
205514          NaN          NaN          NaN          NaN         NaN  \
169351          NaN          NaN          NaN          NaN         NaN   
129869          NaN          NaN          NaN          NaN         NaN   
112855          NaN          NaN          NaN          NaN         NaN   
88950           NaN          NaN          NaN          NaN         NaN   

       Unnamed: 39 Unnamed: 40  
205514         NaN         NaN  
169351         NaN         NaN  
129869         NaN         NaN  
112855         NaN         NaN  
88950          NaN         NaN  

[5 rows x 40 columns]

CONTRIBUTIONTYPE

In [3]:
df.columns

Index(['OBJECTID', 'COMMITTEENAME', 'CANDIDATENAME', 'ELECTIONYEAR',
       'CONTRIBUTORNAME', 'ADDRESS', 'CONTRIBUTORTYPE', 'CONTRIBUTIONTYPE',
       'EMPLOYER', 'EMPLOYERADDRESS', 'AMOUNT', 'DATEOFRECEIPT', 'ADDRESS_ID',
       'XCOORD', 'YCOORD', 'LATITUDE', 'LONGITUDE', 'FULLADDRESS',
       'GIS_LAST_MOD_DTTM', 'WARD', 'Unnamed: 21', 'Unnamed: 22',
       'Unnamed: 23', 'Unnamed: 24', 'Unnamed: 25', 'Unnamed: 26',
       'Unnamed: 27', 'Unnamed: 28', 'Unnamed: 29', 'Unnamed: 30',
       'Unnamed: 31', 'Unnamed: 32', 'Unnamed: 33', 'Unnamed: 34',
       'Unnamed: 35', 'Unnamed: 36', 'Unnamed: 37', 'Unnamed: 38',
       'Unnamed: 39', 'Unnamed: 40'],
      dtype='object')

Drop all columns that have 'Unnamed' in column name

In [4]:
for col in df.columns:
    if 'Unnamed' in col:
        print('Dropping column', col)
        df = df.drop(columns=[col])

Dropping column Unnamed: 21
Dropping column Unnamed: 22
Dropping column Unnamed: 23
Dropping column Unnamed: 24
Dropping column Unnamed: 25
Dropping column Unnamed: 26
Dropping column Unnamed: 27
Dropping column Unnamed: 28
Dropping column Unnamed: 29
Dropping column Unnamed: 30
Dropping column Unnamed: 31
Dropping column Unnamed: 32
Dropping column Unnamed: 33
Dropping column Unnamed: 34
Dropping column Unnamed: 35
Dropping column Unnamed: 36
Dropping column Unnamed: 37
Dropping column Unnamed: 38
Dropping column Unnamed: 39
Dropping column Unnamed: 40


In [5]:
df.columns

Index(['OBJECTID', 'COMMITTEENAME', 'CANDIDATENAME', 'ELECTIONYEAR',
       'CONTRIBUTORNAME', 'ADDRESS', 'CONTRIBUTORTYPE', 'CONTRIBUTIONTYPE',
       'EMPLOYER', 'EMPLOYERADDRESS', 'AMOUNT', 'DATEOFRECEIPT', 'ADDRESS_ID',
       'XCOORD', 'YCOORD', 'LATITUDE', 'LONGITUDE', 'FULLADDRESS',
       'GIS_LAST_MOD_DTTM', 'WARD'],
      dtype='object')

In [6]:
df.dtypes

OBJECTID               int64
COMMITTEENAME         object
CANDIDATENAME         object
ELECTIONYEAR           int64
CONTRIBUTORNAME       object
ADDRESS               object
CONTRIBUTORTYPE       object
CONTRIBUTIONTYPE      object
EMPLOYER              object
EMPLOYERADDRESS       object
AMOUNT               float64
DATEOFRECEIPT         object
ADDRESS_ID           float64
XCOORD               float64
YCOORD               float64
LATITUDE             float64
LONGITUDE            float64
FULLADDRESS           object
GIS_LAST_MOD_DTTM     object
WARD                  object
dtype: object

## Q2

**Points:** 1

**Question:** How many donations (rows) in this dataset are from corporations?

Write a function called `numdoncorp()` the function should take as parameter the name of the file with data and it should return the answer to the question as an int.

In [7]:
import pandas as pd

def numdoncorp(path):
    # YOUR CODE HERE
    df = pd.read_csv(path)
    count_of_corporations = df[df['CONTRIBUTORTYPE']=='Corporation']
    return count_of_corporations['CONTRIBUTORTYPE'].count()
    #raise NotImplementedError()

⬇️⬇️⬇️ Use the cell below to test your code. You can edit this cell or add more cells to test your code ⬇️⬇️⬇️

In [8]:
numdoncorp("Campaign_Financial_Contributions_Sample.csv")

137

There are 137 donations from corporations in this dataset

⬇️⬇️⬇️ The cell below is included for autograding purposes; please do not remove it or change it. ⬇️⬇️⬇️ 

In [9]:
# Autograder cell -- do not remove it!

## Q3

**Points:** 1

**Question:** True or False: The campaign contribution data set is structured data.

1. True
2. False

True

# Part B: cleaning inconsistent records

The next set of questions will ask you to clean the dataframe before computing the desired value.

## Q4

**Points:** 1

**Question:**  What is the minimum donation amount (do not round)? Note that in this an the following questions, any donation amounts of 0 are likely mistakes and are NOT acceptable answers because donors cannot give a $0 donation. Since there are some donations with 0 amount, filter these rows.

Write a function called `findLeastAmount()`. the function should take as parameter the name of the file with data and it should return the exact value and not round off value as the answer.

In [10]:
import pandas as pd

# YOUR CODE HERE
def findLeastAmount(path):
    df = pd.read_csv(path)
    filtered_df = df[df['AMOUNT']!=0]
    return filtered_df['AMOUNT'].min()
#raise NotImplementedError()

Minimum donation amount is .09

⬇️⬇️⬇️ Use the cell below to test your code. You can edit this cell or add more cells to test your code ⬇️⬇️⬇️

In [11]:
findLeastAmount("Campaign_Financial_Contributions_Sample.csv")

0.09

⬇️⬇️⬇️ The cell below is included for autograding purposes; please do not remove it or change it. ⬇️⬇️⬇️ 

In [12]:
# Autograder cell -- do not remove it!

## Q5

<div class="alert alert-warning">Make sure to apply the same cleaning step from Q4</div>

**Points:** 1

**Question:** What is the maximum donation amount?

Write a function called `findMaxAmount()`. the function should take as parameter the name of the file with data and it should return the exact value and not round off value as the answer.

In [13]:
import pandas as pd

# YOUR CODE HERE
def findMaxAmount(path):
    df = pd.read_csv(path)
    filtered_df = df[df['AMOUNT']!=0]
    return filtered_df['AMOUNT'].max()

#raise NotImplementedError()

The max donation amount is $35000

⬇️⬇️⬇️ Use the cell below to test your code. You can edit this cell or add more cells to test your code ⬇️⬇️⬇️

In [14]:
findMaxAmount("Campaign_Financial_Contributions_Sample.csv")

35000.0

⬇️⬇️⬇️ The cell below is included for autograding purposes; please do not remove it or change it. ⬇️⬇️⬇️ 

In [15]:
# Autograder cell -- do not remove it!

## Q6

<div class="alert alert-warning">Make sure to apply the same cleaning step from Q4</div>

**Points:** 1

**Question:** What is the most recent date listed in the column for DATEOFRECEIPT?

Write a function called `findRecentDate()`. the function should take as parameter the name of the file with data and it should return a Python string in the form `YYYY-MM-DD`.

In [16]:
import pandas as pd

# YOUR CODE HERE
def findRecentDate(path):
    df = pd.read_csv(path)
    filtered_df = df[df['AMOUNT']!=0]
    most_recent_date = filtered_df['DATEOFRECEIPT'].max()
    date_cleaned = most_recent_date.split('T')
    return date_cleaned[0]
    
#raise NotImplementedError()

In [17]:
findRecentDate("Campaign_Financial_Contributions_Sample.csv")

'2019-03-13'

In [18]:
# Autograder cell -- do not remove it!

## Q7

<div class="alert alert-warning">Make sure to apply the same cleaning step from the Q4</div>

**Points: 1**

**Question:** How many missing values are there for contributor type (CONTRIBUTORTYPE)?

In [19]:
import pandas as pd

# YOUR CODE HERE
def findMissingValues(path):
    df = pd.read_csv(path)
    filtered_df = df[df['AMOUNT']!=0]
    missing_value_count = filtered_df['CONTRIBUTORTYPE'].isna().sum()
    return missing_value_count
#raise NotImplementedError()


⬇️⬇️⬇️ Use the cell below to test your code. You can edit this cell or add more cells to test your code ⬇️⬇️⬇️

In [20]:
findMissingValues("Campaign_Financial_Contributions_Sample.csv")

1

⬇️⬇️⬇️ The cell below is included for autograding purposes; please do not remove it or change it. ⬇️⬇️⬇️ 

In [21]:
# Autograder cell -- do not remove it!